presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSource.java

```diff
@@ -92,11 +103,35 @@ public boolean isFinished()
     @Override
     public Page getNextPage()
     {
+        boolean supplyRowIDs = recordReader.isColumnPresent(ROW_ID_COLUMN_INDEX);
         try {
             Page page = recordReader.getNextPage();
             if (page == null) {
                 close();
             }
+            else if (supplyRowIDs) { // If we need a row ID block, synthesize it here.
+                // rowNumbers is always the last block in the page
```

Comment:
Would be nice to either create a new class handling this logic (it can be reused in the OrcBatchPageSource) or a new method to make it look cleaner.